src/main.rs

```diff
@@ -0,0 +1,116 @@
+//! Actix Basic
+//!
+//! Tasks defined `rust_actix-basic.pdf`
+//!
+//! - Modeling endpoint behavior base on [this test api](https://jsonplaceholder.typicode.com/api/v1/users)
+//! - This server implementation based on [this basic example](https://github.com/actix/examples/blob/master/basics/src/main.rs)
+
+use ::serde_json;
+use simplelog::*;
+use std::{env, io};
+use validator::Validate;
+
+#[macro_use]
+extern crate actix_web;
+use actix_files as fs;
+use actix_web::http::StatusCode;
+use actix_web::{error, middleware, web, App, Error, HttpResponse, HttpServer, Result};
+
+mod user;
+
+/// favicon handler
+#[get("/favicon.ico")]
+async fn favicon() -> Result<fs::NamedFile> {
+    Ok(fs::NamedFile::open("static/favicon.ico")?)
+}
+
+/// simple index handler
+#[get("/health")]
+async fn health() -> Result<HttpResponse> {
+    Ok(HttpResponse::build(StatusCode::OK)
+        .content_type("text/plain; charset=utf-8")
+        .body("ok"))
+}
+
+/// 404 handler
+async fn p404() -> Result<HttpResponse> {
+    Ok(HttpResponse::new(StatusCode::NOT_FOUND))
+}
+
+/// User GET requests specifying the API version `/api/{version}/users`
+async fn get_users(vers: web::Path<(String,)>) -> Result<HttpResponse> {
+    vers_check(&vers)?;
+
+    Ok(HttpResponse::build(StatusCode::OK)
+        .content_type("application/json; charset=utf-8")
+        .body(include_str!("../static/users.json")))
+    // FIXME: this results in a static file of users, needs to give database values!!
+}
+
+/// User POST requests specifying the API version `/api/{version}/users`
+async fn post_user(
+    vers: web::Path<(String,)>,
+    user: web::Json<user::User>,
+) -> Result<HttpResponse, Error> {
+    vers_check(&vers)?;
+
+    // NOTE: this results in *dummy* response, needs to write to and return database values in production!!
+    let mut new_user = user.into_inner();
+    new_user.id = 11;
+
+    // validation requirements defined in User struct definition, for each element with a #[validate(...)] macro
+    match new_user.validate() {
```

Comment:
Try exploring abstracting validating and just have a one liner in this file that uses the question mark operator.

Reply:
not sure how to do this without an equivalent `if {} else{}` block , as we need to throw a specific error.... What might make this less verbose?